Eventhistory Service (#5600)
owncloud/ocis: services/eventhistory/README.md

> # Eventhistory service
> 
> The `eventhistory` consumes all events from the configured event systems, stores them and allows to retrieve them via an eventid
> 
> ## Consuming
> 
> The `eventhistory` services consumes all events from the configured event sytem. Running it without an event sytem is not possible.
> 
> ## Storing
> 
> The `eventhistory` stores each consumed event in the configured store. Possible stores are:
> - `mem`: basic inmemory store and the default
> - `ocmem`: advanced inmemory store allowing max size
> - `redis`: stores data in a configured redis cluster
> - `etcd`: stores data in a configured etcd cluster
> - `nats-js`: stores data using nats
> - `noop`: stores nothing. Useful for testing. Not recommended in productive enviroments
> 
> ## Retrieving
> 
> Other services can call the `eventhistory` service via a grpc call to retrieve events. The request must contain the eventid that should be retrieved

Reviewer: ```suggestion
# Eventhistory Service

The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows to retrieve them via an eventid.

## Prerequisites

Running the eventhistory service without an event sytem like NATS is not possible.

## Consuming

The `eventhistory` services consumes all events from the configured event sytem.

## Storing

The `eventhistory` stores each consumed event in the configured store. Possible stores are:
- `mem`: basic inmemory store and the default
- `ocmem`: advanced inmemory store allowing max size
- `redis`: stores data in a configured redis cluster
- `etcd`: stores data in a configured etcd cluster
- `nats-js`: stores data using nats
- `noop`: stores nothing. Useful for testing. Not recommended in productive enviroments

## Retrieving

Other services can call the `eventhistory` service via a grpc call to retrieve events. The request must contain the eventid that should be retrieved.
```
The text should include somehow what clears history data. Is it time or retrieval or...?
`Other services can call...` is it only about services or could that include clients?

Author: Yes true. There is an envvar `EVENTHISTORY_RECORD_EXPIRY` which defines how long events survive in the store. But it should be mentioned here.

# Eventhistory Service

The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows to retrieve them via an eventid.

## Prerequisites

Running the eventhistory service without an event sytem like NATS is not possible.

## Consuming

The `eventhistory` services consumes all events from the configured event sytem.

## Storing

> The `eventhistory` stores each consumed event in the configured store. Possible stores are:

Reviewer: ```suggestion
The `eventhistory` service stores each consumed event via the configured store in `EVENTHISTORY_STORE_TYPE`. Possible stores are:
```

>   - `mem`: Basic inmemory store and the default.
>   - `ocmem`: Advanced inmemory store allowing max size.

Reviewer: ```suggestion
  - `mem`: Basic in-memory store and the default.
  - `ocmem`: Advanced in-memory store allowing max size.
```

  - `redis`: Stores data in a configured redis cluster.
  - `etcd`: Stores data in a configured etcd cluster.

>   - `nats-js`: Stores data using nats.

Reviewer: Though I heard about redis and etcd, I never heard of `nats-js`. Maybe to add a link for more reading?

Author: It's `nats jetstream`. The same technology we are using as eventbus. It also provides key-value store functionality. I added a link.

  - `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.

> Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.

Reviewer: ```suggestion
1.   Note that in-memory stores are by nature not reboot persistant.
2.   Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
```
Added a presistance info and made a numbered list of items.

## Retrieving

> Other services can call the `eventhistory` service via a grpc call to retrieve events. The request must contain the eventid that should be retrieved.

Reviewer: > The request must contain the eventid that should be retrieved.

How does one know a eventID?

Author: One cannot. As said it is only other services that can call eventhistory service. Other service will automatically get the id when consuming events.

# Eventhistory Service

> The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows to retrieve them via an eventid.

Reviewer: ```suggestion
The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows other services to retrieve them via an eventid.
```

## Prerequisites

Running the eventhistory service without an event sytem like NATS is not possible.

## Consuming

The `eventhistory` services consumes all events from the configured event sytem.

## Storing

The `eventhistory` service stores each consumed event via the configured store in `EVENTHISTORY_STORE_TYPE`. Possible stores are:

>   - `mem`: Basic in-memory store and the default.
>   - `ocmem`: Advanced in-memory store allowing max size.
>   - `redis`: Stores data in a configured redis cluster.
>   - `etcd`: Stores data in a configured etcd cluster.

Reviewer: ```suggestion
  -   `mem`: Basic in-memory store and the default.
  -   `ocmem`: Advanced in-memory store allowing max size.
  -   `redis`: Stores data in a configured redis cluster.
  -   `etcd`: Stores data in a configured etcd cluster.
```
for codacy

>   - `nats-js`: Stores data using key-value-store feature of nats jetstream. https://docs.nats.io/nats-concepts/jetstream/key-value-store

Reviewer: ```suggestion
  - `nats-js`: Stores data using key-value-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/key-value-store)
```

>   - `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.

Reviewer: ```suggestion
  -   `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
```
for codacy

> 1.   Note that in-memory stores are by nature not reboot persistant.

Reviewer: ```suggestion
1.  Note that in-memory stores are by nature not reboot persistant.
```

> 2.   Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.

Reviewer: ```suggestion
2.   Though usually not necessary, a database name and a database table can be configured for event stores that are not of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store applies.
3.   Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
```
